### PR TITLE
pass non-zero timeout value to spdm IO.

### PR DESCRIPTION
**Pull request by qizhangz**
Fix: #272
Fix: #494
Fix: #20 

Signed-off-by: Qi Zhang <qi1.zhang@intel.com>

### CODE REVIEW

File: include/library/spdm_lib_config.h [hunk 1]
```diff
@@ -87,4 +87,13 @@
 #define LIBSPDM_ENABLE_CAPABILITY_KEY_EX_CAP 1
 #define LIBSPDM_ENABLE_CAPABILITY_PSK_EX_CAP 1
 
+/* Worst case round-trip transport timing in microseconds */
+#define LIBSPDM_RTT_VALUE               10000
```
Comment (jyao1) [inline include/library/spdm_lib_config.h:91]: Please remove it and expose set_data() API.

File: include/library/spdm_lib_config.h [hunk 2]
```diff
@@ -87,4 +87,13 @@
...
+
+/**
+  2^CT is maximum amount of time in microseconds that the endpoint has
+  to provide any response requiring cryptographic processing
+**/
+#define LIBSPDM_CT_EXPONENT_VALUE       20
```
Comment (jyao1) [inline include/library/spdm_lib_config.h:97]: Please remove it and expose set_data() API.

**Comment by jyao1** [comment]
looks good on other part.

File: include/library/spdm_lib_config.h (hunk 1 above)
Comment (steven-bellock) [reply include/library/spdm_lib_config.h:91]: For both RTT and CT I don't think having them as `#define`s provides much optimization. I think it's fine to remove them from `spdm_lib_config.h` and just use `set_data`.

File: library/spdm_requester_lib/libspdm_req_challenge.c [hunk 3]
```diff
@@ -113,15 +113,19 @@ return_status try_spdm_challenge(IN void *context, IN uint8_t slot_id,
 
     status = spdm_send_spdm_request(spdm_context, NULL,
                     sizeof(spdm_request), &spdm_request);
-    if (RETURN_ERROR(status)) {
+    if (status == RETURN_TIMEOUT) {
+        return status;
+    } else if (RETURN_ERROR(status)) {
         return RETURN_DEVICE_ERROR;
     }
```
Comment (steven-bellock) [inline library/spdm_requester_lib/libspdm_req_challenge.c:121]: This file, and lots of other requests, needs `spdm_context->crypto_request = TRUE;`

**Review by steven-bellock — CHANGES_REQUESTED** [review]
Lots of files, like

File: library/spdm_requester_lib/libspdm_req_challenge.c (hunk 3 above)
Comment (qizhangz) [reply library/spdm_requester_lib/libspdm_req_challenge.c:121]: spdm_context->crypto_request is set to TRUE in libspdm_init_context(). To GET_CAPABILITIES , GET_VERSION , or NEGOTIATE_ALGORITHMS request message, it is set to FALSE.

Comment (steven-bellock) [reply library/spdm_requester_lib/libspdm_req_challenge.c:121]: How does the sequence `NEGOTIATE_ALGORITHMS (crypto_request == FALSE)` -> `GET_DIGESTS (crypto_request == TRUE)` occur then? There's no `libspdm_init_context` in between the two requests.

Comment (qizhangz) [reply library/spdm_requester_lib/libspdm_req_challenge.c:121]: I got your point. Will update the change  for spdm_context->crypto_request = TRUE.

File: library/spdm_responder_lib/libspdm_rsp_psk_exchange.c [hunk 4]
```diff
@@ -359,6 +360,10 @@ return_status spdm_get_response_psk_exchange(IN void *context,
 
         spdm_set_session_state(spdm_context, session_id,
                        LIBSPDM_SESSION_STATE_ESTABLISHED);
+        result = init_watchdog(spdm_context->local_context.heartbeat_period);
```
Comment (jyao1) [inline library/spdm_responder_lib/libspdm_rsp_psk_exchange.c:363]: why we need special handling here?
Can we move it together with libspdm_rsp_receive_send.c ?

File: library/spdm_requester_lib/libspdm_req_challenge.c [hunk 5]
```diff
@@ -112,15 +112,19 @@ return_status try_spdm_challenge(IN void *context, IN uint8_t slot_id,
...
 
     spdm_response_size = sizeof(spdm_response);
     zero_mem(&spdm_response, sizeof(spdm_response));
     status = spdm_receive_spdm_response(
         spdm_context, NULL, &spdm_response_size, &spdm_response);
-    if (RETURN_ERROR(status)) {
+    if (status == RETURN_TIMEOUT) {
+        return status;
+    } else if (RETURN_ERROR(status)) {
```
Comment (jyao1) [inline library/spdm_requester_lib/libspdm_req_challenge.c:127]: Do we really want to use this pattern?
Can we just make it simple:
```
if (RETURN_ERROR(status)) {
        return status;
}
```

Or, if we really want to do some special handling, why not do it inside of spdm_receive_spdm_response() ?

File: include/hal/library/platform_lib.h [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+void libspdm_sleep(uint64_t milliseconds);
+
+/**
+  If no heartbeat arrives in microseconds, the watchdog timeout event
+  should terminate the session.
+
+  @param[in] microseconds     heartbeat period, in microseconds.
+
+**/
+boolean init_watchdog(uint64_t microseconds);
```
Comment (jyao1) [inline include/hal/library/platform_lib.h:25]: I think we need rename it to start_watchdog(), and we need add stop_watchdog() and always call stop_watchdog() when the session is ended.

File: library/spdm_responder_lib/libspdm_rsp_psk_exchange.c [hunk 7]
```diff
@@ -359,6 +360,11 @@ return_status spdm_get_response_psk_exchange(IN void *context,
 
         spdm_set_session_state(spdm_context, session_id,
                        LIBSPDM_SESSION_STATE_ESTABLISHED);
+        result = start_watchdog(&spdm_context->watchdog_context, session_id,
```
Comment (steven-bellock) [inline library/spdm_responder_lib/libspdm_rsp_psk_exchange.c:363]: Heartbeats are only active in the application phase, so this should be in `PSK_FINISH_RSP`. `FINISH_RSP` needs it as well.

File: library/spdm_responder_lib/libspdm_rsp_psk_exchange.c [hunk 8]
```diff
@@ -359,6 +360,11 @@ return_status spdm_get_response_psk_exchange(IN void *context,
...
+            spdm_context->local_context.heartbeat_period);
```
Comment (steven-bellock) [inline library/spdm_responder_lib/libspdm_rsp_psk_exchange.c:364]: `local_context.heartbeat_period` is in units of seconds and `start_watchdog` uses units of microseconds.

**Review by steven-bellock — CHANGES_REQUESTED** [review]

**Comment by steven-bellock** [comment]
@jyao1 This pull request inspired https://github.com/DMTF/Security-TF/issues/1795

File: include/hal/library/platform_lib.h [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+void libspdm_sleep(uint64_t milliseconds);
+
+/**
+  If no heartbeat arrives in seconds, the watchdog timeout event
+  should terminate the session.
+
+  @param[in] seconds     heartbeat period, in seconds.
+
+**/
+boolean libspdm_start_watchdog(void *context, uint32_t session_id, uint64_t seconds);
```
Comment (jyao1) [inline include/hal/library/platform_lib.h:32]: What is "void *context" ?

File: include/internal/libspdm_common_lib.h [hunk 10]
```diff
@@ -377,6 +378,8 @@ typedef struct {
     // or secured SPDM transport message.
     //
     uint8_t request_response[LIBSPDM_MAX_MESSAGE_BUFFER_SIZE];
+
+    watchdog_context_t watchdog_context;
```
Comment (jyao1) [inline include/internal/libspdm_common_lib.h:382]: why you need add watchdog_context_t ?

we already have heartbeat_period.

**Comment by jyao1** [comment]
Looks good to me now.

**Comment by jyao1** [comment]
@steven-bellock, since hearbeat 0 is handled in #549, can we merge this at first?

**Review by jyao1 — APPROVED** [review]

**Comment by steven-bellock** [comment]
Let's discuss this pull request at Mondays meeting. I'm still thinking about the watchdog.

**Comment by steven-bellock** [comment]
@jyao1 let's discuss this one at Mondays meeting. I still have questions about the watchdog.

**Review by steven-bellock — APPROVED** [review]
We can approve for now, but there will need to be a follow-up pull request to address issues such as number of watchdog timers available and resetting the watchdog.

**Comment by qizhangz** [comment]
rebased to resolve the conflict.

**Comment by jyao1** [comment]
@qizhangz, CI still fails.

**Comment by qizhangz** [comment]
CI passed with coding style issue fixed.

**Pull request merged**